### DISCUSSION TITLE
Bind Repeater Rows to deeply nested json array

### DISCUSSION THREAD

**Discussion by MaxBernstetter** [Budibase/budibase#9248] [Help] [answered]
``Hi, 

I need to recursibvely map data from a nested json object (queried from a PSQL-Datasource) to repeater blocks. 

A data object looks like this:
k1: v1
k2: v2
k3: 
[ 
  {
     k31: v31, 
     k32: v32, 
     k33: [ {
               k331: v331 
               k332: v332 
               k333: v333 
               },
               {
               k331: v334
               k332: v335 
               k333: v336 
               }
             ]
   }
  {
   ...
  }
]

---> Note the double nested array within the object. 

Concrete aim: 
I can generate a repeater block for the outer array. BB lets me use the 'k3' key as data for a repeater. Now, I want to do the same, but one level deeper in the object. I want to create a card-object for each object having the k33X key-value pairs. 

Many thanks!

**Comment by melohagan** [comment, marked as answer]
Hey @MaxBernstetter 

This one is quite nasty because you needs to convert objects of key/value pairs into an iterable array.

Here's something to get you started. Add the following to the transformer of a custom Postgres Query:

```js
const test = {
	k1: "v1",
	k3: [
		{
			k31: "v31",
			k33: [
				{
					k331: "v331",
					k332: "v332",
					k333: "v333"
				},
				{
					k331: "v334",
					k332: "v335",
					k333: "v336"
				}
			]
		}
	]
}

let result = []
for (let obj of test["k3"][0]["k33"]) {
	for (let entry of Object.entries(obj)) {
		result.push(entry)
	}
}
return result
```

I hard coded the test data here, but this may simply be **data** as returned from your SQL query.

Result:
![Screenshot 2023-01-10 at 09 22 00](https://user-images.githubusercontent.com/101575380/211511807-227495a9-b01e-46e4-87be-6b0e95dee394.png)

Use `.0` and .1` for the key and value.
![Screenshot 2023-01-10 at 09 22 22](https://user-images.githubusercontent.com/101575380/211511883-363a5c57-f7c8-4d8b-b343-b6b2ad1be07c.png)

---

This is really not ideal though. I personally think that you should try to store the data in a more structured manner in the first place if that's an option.

**Comment by MaxBernstetter** [comment]
Hi @melohagan 

many thanks for your help! :) I will give it a try. I implemented it in order to reduce the number of necessary data-source calls to 1. I was hoping to gain screen loading-speed from this.

**Reply by melohagan** [reply]
Well you have to remember too that a lot of JavaScript looping could definitely slow down screen loading.

Usually you want to offload the 'heavy lifting' server side/DB side before pulling the data into your UI.
